Improve login by magiclink

FabriceIRANKUNDA opened the pull request:
## Summary
Updated link sent via email 
Improved magic link to be used once and before expiration

## Checklist
- [ ] Pre-push checks pass `npm run check`
  - [ ] Linter passing via `npm run lint`
  - [ ] Unit & Integration tests passing via `npm run test:packages`
  - [ ] Docker build process passing via `npm run build-client`
- [ ] If this PR is still a WIP, convert to a draft 
- [ ] When this PR is ready, mark it as "Ready for review"
- [ ] Changes have been manually QA'd
- [ ] Changes reviewed by at least 2 approved reviewers

## References

_References to pertaining issue(s)_

## QA Steps

1. `git checkout pr_branch_name`
2. `npm install`
3. `npm run dev-reinit`
4. `npm run dev`

List any additional steps required to QA the changes of this PR, as well as any supplemental images or videos.

## Reviewers

_Reviewers for this PR_

barankyle commented:
@FabriceIRANKUNDA I fixed some issues I'd found. The major one was that, seemingly for a while (well before you started working on this), submitting an email/sms was calling AuthService.addConnectionByEmail/BySMS, which is not correct. Our authentication system supports have multiple identity providers associated with a single user, but we've never really implemented it to date. addConnectionBy<Service> is only used when adding something to an existing User, not when logging in from a Guest account.

I also changed up how loginUserMagicLink works. I have it doing a GET on the 'login' service, which is what was handling the magiclink logins before, only now it's being called from the front-end rather than been hit directly by the browser. This returns a JWT generated from the associated identity-provider, which is then passed into loginUserByJWT(), which will save the JWT into localStorage and actually log the user in. the loadUserData() call you were trying to do doesn't persist over refreshes, since the client is set up to use the JWT stored in localStorage to authorize all of the calls to the API.

I made 'login-token'.FIND disallowed externally again, as that was leaking the login_token table. 'login'.GET should be more secure about only providing a valid JWT when a valid login token is passed to it. I also got rid of what you'd written for login-token.FIND since it wasn't needed anymore; the only lookup we do in login.GET is a findOne using the model.

FabriceIRANKUNDA commented:
> @FabriceIRANKUNDA I fixed some issues I'd found. The major one was that, seemingly for a while (well before you started working on this), submitting an email/sms was calling AuthService.addConnectionByEmail/BySMS, which is not correct. Our authentication system supports have multiple identity providers associated with a single user, but we've never really implemented it to date. addConnectionBy(Service) is only used when adding something to an existing User, not when logging in from a Guest account.
> 
> I also changed up how loginUserMagicLink works. I have it doing a GET on the 'login' service, which is what was handling the magiclink logins before, only now it's being called from the front-end rather than been hit directly by the browser. This returns a JWT generated from the associated identity-provider, which is then passed into loginUserByJWT(), which will save the JWT into localStorage and actually log the user in. the loadUserData() call you were trying to do doesn't persist over refreshes, since the client is set up to use the JWT stored in localStorage to authorize all of the calls to the API.
> 
> I made 'login-token'.FIND disallowed externally again, as that was leaking the login_token table. 'login'.GET should be more secure about only providing a valid JWT when a valid login token is passed to it. I also got rid of what you'd written for login-token.FIND since it wasn't needed anymore; the only lookup we do in login.GET is a findOne using the model.

Awesome, thanks.
